src/irmin-pack/inode.ml

```diff
@@ -1414,13 +1444,44 @@ struct
   type key = Key.t
   type value = Inter.Val.t
 
+  exception Invalid_depth = Inter.Raw.Invalid_depth
+
   let mem t k = Pack.mem t k
+  let pp_value = Irmin.Type.pp Inter.Raw.t
+
+  let err_invalid_depth expected got v =
+    raise (Invalid_depth { expected; got; v })
+
+  let pp_invalid_depth ppf (expected, got, v) =
+    Fmt.pf ppf "Invalid depth: got %a, expecting %d (%a)"
+      Fmt.(Dump.option int)
+      got expected pp_value v
+
+  let check_depth v e g =
+    match g with
+    | None -> if e = 0 then err_invalid_depth e g v
+    | Some depth -> if depth <> e then err_invalid_depth e g v
+
+  let check_depth_opt ~init_depth ~depth v =
+    match v with
+    | None -> ()
+    | Some v -> (
+        match (init_depth, Inter.Raw.depth v) with
+        | None, None -> ()
+        | None, _ -> assert false
+        | Some d0, Some d -> check_depth v depth (Some (d0 + d))
```

Comment:
```suggestion
        | Some d0, Some d -> check_depth v depth (d0 + d)
```
this is the only place where `check_depth` is called, you can refactor to remove the option